dali/imgcodec/decoders/nvjpeg.h

```diff
@@ -0,0 +1,112 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_IMGCODEC_DECODERS_NVJPEG_H_
+#define DALI_IMGCODEC_DECODERS_NVJPEG_H_
+
+#include <nvjpeg.h>
+#include <memory>
+#include <vector>
+#include "dali/core/cuda_event.h"
+#include "dali/core/cuda_stream_pool.h"
+#include "dali/core/dev_buffer.h"
+#include "dali/imgcodec/decoders/decoder_parallel_impl.h"
+#include "dali/operators/decoder/nvjpeg/nvjpeg_memory.h"
+
+namespace dali {
+namespace imgcodec {
+
+class DLL_PUBLIC NvJpegDecoderInstance : public BatchParallelDecoderImpl {
+ public:
+  NvJpegDecoderInstance(int device_id, ThreadPool *tp);
+
+  DecodeResult DecodeImplTask(int thread_idx,
+                              cudaStream_t stream,
+                              SampleView<GPUBackend> out,
+                              ImageSource *in,
+                              DecodeParams opts,
+                              const ROI &roi) override;
+  ~NvJpegDecoderInstance();
+
+  void SetParam(const char *name, const any &value) override;
+  any GetParam(const char *name) const override;
+
+ private:
+  nvjpegHandle_t nvjpeg_handle_;
+
+  size_t device_memory_padding_;
+  size_t host_memory_padding_;
+  nvjpegDevAllocator_t device_allocator_;
+  nvjpegPinnedAllocator_t pinned_allocator_;
+
+  struct DecoderData {
+    nvjpegJpegDecoder_t decoder = nullptr;
+    nvjpegJpegState_t state = nullptr;
+  };
+
+  struct PerThreadResources {
+    DecoderData decoder_data;
+
+    nvjpegBufferDevice_t device_buffer;
+    nvjpegBufferPinned_t pinned_buffer;
+
+    nvjpegJpegStream_t jpeg_stream;
+    CUDAStreamLease stream;
+    CUDAEvent decode_event;
+
+    PerThreadResources(nvjpegHandle_t, nvjpegDevAllocator_t*, nvjpegPinnedAllocator_t*,
+                       int device_id);
+    PerThreadResources(PerThreadResources&&);
+    ~PerThreadResources();
+  };
+
+  std::vector<PerThreadResources> resources_;
+
+  struct DecodingContext {
+    PerThreadResources& resources;
+
+    nvjpegDecodeParams_t params;
+    nvjpegChromaSubsampling_t subsampling;
+
+    TensorShape<> shape;
+  };
+
+  void ParseJpegSample(ImageSource& in, DecodeParams opts, DecodingContext& ctx);
+  void DecodeJpegSample(ImageSource& in, uint8_t *out, DecodeParams opts, DecodingContext &ctx);
+};
+
+class NvJpegDecoderFactory : public ImageDecoderFactory {
+ public:
+  ImageDecoderProperties GetProperties() const override {
+    ImageDecoderProperties props;
+    props.supports_partial_decoding = false;
```

Comment:
add TODO here

Reply:
Wait, I feel like I'm receiving conflicting instructions here... Few review comments before @mzient told me to remove a similar comment (https://github.com/NVIDIA/DALI/pull/4178#discussion_r949161861)

Reply:
nvJPEG does have partial decoding APIs, so we will implement it soon. I think the TODO makes sense here. I'll sync offline with mzient about the previous comment.